src/main/java/io/cdap/plugin/gcp/firestore/source/QueryDocumentSnapshotToRecordTransformer.java

```diff
@@ -0,0 +1,223 @@
+          builder.set(field.getName(), "");
+        } else {
+          builder.set(field.getName(), value);
+        }
+      }
+    }
+    return builder.build();
+  }
+
+  private Object extractValue(String fieldName, Object object, Schema schema) {
+    //if (object == null || object instanceof BsonUndefined) {
+    if (object == null) {
+      return null;
+    }
+
+    Schema.LogicalType fieldLogicalType = schema.getLogicalType();
+    if (fieldLogicalType != null) {
+      switch (fieldLogicalType) {
+        case TIMESTAMP_MILLIS:
+          ensureTypeValid(fieldName, object, Date.class);
+          Instant instant = ((Date) object).toInstant();
+          long millis = TimeUnit.SECONDS.toMillis(instant.getEpochSecond());
+          return Math.addExact(millis, TimeUnit.NANOSECONDS.toMillis(instant.getNano()));
+        case TIMESTAMP_MICROS:
+          ensureTypeValid(fieldName, object, Date.class);
+          Instant dateInstant = ((Date) object).toInstant();
+          long micros = TimeUnit.SECONDS.toMicros(dateInstant.getEpochSecond());
+          return Math.addExact(micros, TimeUnit.NANOSECONDS.toMicros(dateInstant.getNano()));
+          /*
+        case DECIMAL:
+          ensureTypeValid(fieldName, object, Decimal128.class);
+          return extractDecimal(fieldName, (Decimal128) object, schema);
+          */
+        default:
+          throw new UnexpectedFormatException(String.format("Field '%s' is of unsupported type '%s'",
+            fieldName, fieldLogicalType.name().toLowerCase()));
+      }
+    }
+
+    Schema.Type fieldType = schema.getType();
+    switch (fieldType) {
+      case BOOLEAN:
+        ensureTypeValid(fieldName, object, Boolean.class);
+        return object;
+      case INT:
+        ensureTypeValid(fieldName, object, Integer.class);
+        return object;
+      case DOUBLE:
+        ensureTypeValid(fieldName, object, Double.class);
+        return object;
+      case BYTES:
+        /*
+        ensureTypeValid(fieldName, object, ObjectId.class, byte[].class);
+        if (object instanceof ObjectId) {
+          // ObjectId can be read as a bytes. Exactly 12 bytes are required
+          ByteBuffer buffer = ByteBuffer.allocate(12);
+          ObjectId objectId = (ObjectId) object;
+          objectId.putToByteBuffer(buffer);
+          buffer.rewind();
+          return buffer;
+        }
+        */
+        return object;
+      case LONG:
+        ensureTypeValid(fieldName, object, Long.class);
+        return object;
+      case STRING:
+        /*
+        ensureTypeValid(fieldName, object, ObjectId.class, String.class);
+        if (object instanceof ObjectId) {
+          // ObjectId can be read as a string
+          return object.toString();
+        }
+        */
+        return object;
+      /*
+      case MAP:
+        ensureTypeValid(fieldName, object, BSONObject.class);
+        return convertMap(fieldName, (BSONObject) object, schema);
```

Comment:
why do we not handle these cases now?

Reply:
We are not supporting this in this version